consensus/src/tests/bullshark_tests.rs

```diff
@@ -0,0 +1,346 @@
+        store,
+        rx_waiter,
+        tx_primary,
+        tx_output,
+        tusk,
+    );
+    tokio::spawn(async move { while rx_primary.recv().await.is_some() {} });
+
+    // Feed all certificates to the consensus. Only the last certificate should trigger
+    // commits, so the task should not block.
+    while let Some(certificate) = certificates.pop_front() {
+        tx_waiter.send(certificate).await.unwrap();
+    }
+
+    // Ensure the first 4 ordered certificates are from round 1 (they are the parents of the committed
+    // leader); then the leader's certificate should be committed.
+    for _ in 1..=4 {
+        let output = rx_output.recv().await.unwrap();
+        assert_eq!(output.certificate.round(), 1);
+    }
+    let output = rx_output.recv().await.unwrap();
+    assert_eq!(output.certificate.round(), 2);
+}
+
+// Run for 8 dag rounds with one dead node node (that is not a leader). We should commit the leaders of
+// rounds 2, 4, and 6.
+#[tokio::test]
+async fn dead_node() {
+    // Make the certificates.
+    let mut keys: Vec<_> = test_utils::keys(None)
+        .into_iter()
+        .map(|kp| kp.public().clone())
+        .collect();
+    keys.sort(); // Ensure we don't remove one of the leaders.
+    let _ = keys.pop().unwrap();
+
+    let genesis = Certificate::genesis(&mock_committee(&keys[..]))
+        .iter()
+        .map(|x| x.digest())
+        .collect::<BTreeSet<_>>();
+
+    let (mut certificates, _) = test_utils::make_optimal_certificates(1..=9, &genesis, &keys);
+
+    // Spawn the consensus engine and sink the primary channel.
+    let (tx_waiter, rx_waiter) = channel(1);
+    let (tx_primary, mut rx_primary) = channel(1);
+    let (tx_output, mut rx_output) = channel(1);
+    let store_path = test_utils::temp_dir();
+    let store = make_consensus_store(&store_path);
+    let tusk = Bullshark {
+        committee: mock_committee(&keys[..]),
+        store: store.clone(),
+        gc_depth: 50,
+    };
+    Consensus::spawn(
+        mock_committee(&keys[..]),
+        store,
+        rx_waiter,
+        tx_primary,
+        tx_output,
+        tusk,
+    );
+    tokio::spawn(async move { while rx_primary.recv().await.is_some() {} });
+
+    // Feed all certificates to the consensus.
+    tokio::spawn(async move {
+        while let Some(certificate) = certificates.pop_front() {
+            tx_waiter.send(certificate).await.unwrap();
+        }
+    });
+
+    // We should commit 4 leaders (rounds 2, 4, 6, and 8).
+    for i in 1..=21 {
+        let output = rx_output.recv().await.unwrap();
+        let expected = ((i - 1) / keys.len() as u64) + 1;
+        assert_eq!(output.certificate.round(), expected);
+    }
+    let output = rx_output.recv().await.unwrap();
+    assert_eq!(output.certificate.round(), 8);
+}
+
+// Run for 5 dag rounds. The leaders of round 2 does not have enough support, but the leader of
+// round 4 does. The leader of rounds 2 and 4 should thus be committed.
```

Comment:
This scenario is not always true (in this deterministic test it is). Should we add an explanation that leader 2 is part of leader's 4 history?

Reply:
good point, we need to ensure that the leader of round 4 has a path to the leader of round 2. It is the case in the test, but I should update the comment.